aten/src/ATen/detail/BackendDispatch.cpp

```diff
@@ -0,0 +1,54 @@
+#include <ATen/detail/BackendDispatch.h>
+
+#include <cstdint>
+
+#include <c10/core/DispatchKey.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+
+using c10::impl::tls_set_dispatch_key_excluded;
+using c10::impl::tls_set_dispatch_key_included;
+
+namespace at {
+namespace detail {
+namespace {
+
+void excludeBackendDispatchKeys(bool desired_state) {
+  // Add all backend keys except meta to the TLS exclude list. This will force
+  // the dispatcher to suspend all tensor operations.
+  auto end = static_cast<std::uint8_t>(DispatchKey::EndOfBackendKeys);
+  for (std::uint8_t i = 1; i <= end; i++) {
+    auto dk = static_cast<DispatchKey>(i);
+    if (dk == DispatchKey::Meta) {
+      continue;
+    }
+    tls_set_dispatch_key_excluded(dk, desired_state);
+  }
+
+  // In addition to the backend ops, we also have to stop the backend selection
+  // logic; otherwise, some operations such as `empty` can escape the exclusion
+  // list.
+  tls_set_dispatch_key_excluded(DispatchKey::BackendSelect, desired_state);
+}
+
+} // namespace
+
+void suspendBackendDispatch() {
+  excludeBackendDispatchKeys(/*desired_state*/true);
+
+  // Now that we have excluded all backends from the dispatcher, enable meta
+  // dispatch key which effectively diverts all backend ops to meta.
+  tls_set_dispatch_key_included(DispatchKey::Meta, /*desired_state*/true);
```

Comment:
curious, do you know what's the reason that we need separate excluded and included lists?

Reply:
In fact we only need to exclude dispatch keys that have a larger numerical value than `Meta`. However I personally find that approach fragile. In case, for some reason, the `DispatchKey` enum entries get reordered in the future that "invariant" would fail. For that reason, to be on the safe side, I exclude all backend dispatch keys by putting them to the exclusion list. Since `Meta` won't be in the requested `DispatchKeySet` by the caller (most of the time), I also make sure to put in the include list.

To learn more how these two (thread-local) lists are used, check out `DispatchKeyExtractor` (here)[https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/dispatch/DispatchKeyExtractor.h#L45]. Both lists together are used to intercept the dispatch logic of the `Dispatcher`.